In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [5]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
file_path = Path("Resources/lending_data.csv")
df_lending = pd.read_csv(file_path)
df_lending

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

In [6]:
# Review the DataFrame
df_lending.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


In [7]:
df_lending.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = df_lending["loan_status"]
# Separate the X variable, the features
X = df_lending.drop(columns="loan_status")

In [12]:
# Review the y variable Series
y.sample(16)

7194     0
51789    0
18992    0
269      0
45763    0
62802    0
47494    0
22584    0
12367    0
63932    0
74217    0
27918    0
23721    0
59228    0
3728     0
48317    0
Name: loan_status, dtype: int64

In [13]:
# Review the X variable DataFrame
X.head(23)

loan_size  interest_rate  borrower_income  debt_to_income  \
0     10700.0          7.672            52800        0.431818   
1      8400.0          6.692            43600        0.311927   
2      9000.0          6.963            46100        0.349241   
3     10700.0          7.664            52700        0.430740   
4     10800.0          7.698            53000        0.433962   
5     10100.0          7.438            50600        0.407115   
6     10300.0          7.490            51100        0.412916   
7      8800.0          6.857            45100        0.334812   
8      9300.0          7.096            47400        0.367089   
9      9700.0          7.248            48800        0.385246   
10     9300.0          7.085            47300        0.365751   
11     9200.0          7.015            46600        0.356223   
12     9200.0          7.043            46900        0.360341   
13     9400.0          7.101            47400        0.367089   
14     8400.0          6.703            43700        0.313501   
15     9400.0          7.136            47700        0.371069   
16     8900.0          6.897            45500        0.340659   
17     8300.0          6.648            43200        0.305556   
18     8000.0          6.539            42100        0.287411   
19    10200.0          7.460            50800        0.409449   
20     5200.0          5.326            30700        0.022801   
21     9300.0          7.068            47100        0.363057   
22     7300.0          6.225            39200        0.234694   

    num_of_accounts  derogatory_marks  total_debt  
0                 5                 1       22800  
1                 3                 0       13600  
2                 3                 0       16100  
3                 5                 1       22700  
4                 5                 1       23000  
5                 4                 1       20600  
6                 4                 1       21100  
7                 3                 0       15100  
8                 3                 0       17400  
9                 4                 0       18800  
10                3                 0       17300  
11                3                 0       16600  
12                3                 0       16900  
13                3                 0       17400  
14                3                 0       13700  
15                3                 0       17700  
16                3                 0       15500  
17                2                 0       13200  
18                2                 0       12100  
19                4                 1       20800  
20                0                 0         700  
21                3                 0       17100  
22                2                 0        9200

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [16]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [18]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier

LogisticRegression(random_state=1)

In [19]:
# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [21]:
# Make a prediction using the testing data
predicted_y_values = classifier.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [22]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predicted_y_values)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [24]:
# Print the classification report for the model
print(classification_report(y_test, predicted_y_values))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model scores well on predicting the values for the healthy loans ("0") with an accuracy rate of 100%, but less so for the unhealthy loans ("1"), although still with a high degree of accuracy at 87%.  In other words, this is accurate.

---